src/WinRT.Runtime/IWinRTObject.net5.cs

```diff
@@ -1,217 +1,298 @@
+                if (!QueryInterfaceCache.TryAdd(interfaceType, typedObjRef))
+                {
+                    typedObjRef.Dispose();
+                }
+                return true;
+            }
+        }
+
+#if NET8_0_OR_GREATER
+        unsafe bool LookupGeneratedVTableInfo(RuntimeTypeHandle interfaceType, [NotNullWhen(true)] out IIUnknownCacheStrategy.TableInfo? result, out int qiResult)
+        {
+            result = null;
+            qiResult = 0;
+            if (AdditionalTypeData.TryGetValue(interfaceType, out object value))
+            {
+                if (value is IIUnknownCacheStrategy.TableInfo tableInfo)
+                {
+                    result = tableInfo;
+                    return true;
+                }
+                return false;
+            }
+
+            if (StrategyBasedComWrappers.DefaultIUnknownInterfaceDetailsStrategy.GetIUnknownDerivedDetails(interfaceType) is IIUnknownDerivedDetails details)
+            {
+                qiResult = NativeObject.TryAs(details.Iid, out ObjectReference<IUnknownVftbl> objRef);
+                if (qiResult < 0)
+                    return false;
+                var obj = (void***)objRef.ThisPtr;
+                result = new IIUnknownCacheStrategy.TableInfo()
+                {
+                    ThisPtr = obj,
+                    Table = *obj,
+                    ManagedType = details.Implementation.TypeHandle
+                };
+
+                if (!AdditionalTypeData.TryAdd(interfaceType, result))
+                {
+                    bool found = AdditionalTypeData.TryGetValue(interfaceType, out object newInfo);
+                    System.Diagnostics.Debug.Assert(found);
+                    result = (IIUnknownCacheStrategy.TableInfo)newInfo;
+                    objRef.Dispose();
+                }
+
+                return true;
+            }
+            return false;
+        }
+#endif
+
+        RuntimeTypeHandle IDynamicInterfaceCastable.GetInterfaceImplementation(RuntimeTypeHandle interfaceType)
+        {
+#if NET8_0_OR_GREATER
+            if (AdditionalTypeData.TryGetValue(interfaceType, out object value) && value is IIUnknownCacheStrategy.TableInfo tableInfo)
+            {
+                return tableInfo.ManagedType;
+            }
+#endif
+            var type = Type.GetTypeFromHandle(interfaceType);
+            var helperType = type.GetHelperType();
+            if (helperType.IsInterface)
+                return helperType.TypeHandle;
+            return default;
+        }
+
+#if NET8_0_OR_GREATER
+        unsafe VirtualMethodTableInfo IUnmanagedVirtualMethodTableProvider.GetVirtualMethodTableInfoForKey(Type type)
+        {
+            if (!LookupGeneratedVTableInfo(type.TypeHandle, out IIUnknownCacheStrategy.TableInfo? result, out int qiHResult))
+            {
+                Marshal.ThrowExceptionForHR(qiHResult);
+            }
+
+            return new((void*)NativeObject.ThisPtr, result.Value.Table);

```

Comment:
Should this be the ptr from the table / cache or is the `Ptr` that may not be QIed for the interface already fine?